rust/lance/src/dataset/write/merge_insert.rs

```diff
@@ -4520,4 +4535,256 @@ MergeInsert: on=[id], when_matched=UpdateAll, when_not_matched=InsertAll, when_n
+        let value_index = indices.iter().find(|idx| idx.name == "value_idx").unwrap();
+        let vec_index = indices.iter().find(|idx| idx.name == "vec_idx").unwrap();
+
+        assert_eq!(value_index.fragment_bitmap.as_ref().unwrap().len(), 2);
+        assert!(value_index.fragment_bitmap.as_ref().unwrap().contains(0));
+        assert!(value_index.fragment_bitmap.as_ref().unwrap().contains(1));
+        assert_eq!(vec_index.fragment_bitmap.as_ref().unwrap().len(), 2);
+        assert!(vec_index.fragment_bitmap.as_ref().unwrap().contains(0));
+        assert!(vec_index.fragment_bitmap.as_ref().unwrap().contains(1));
+
+        // update keys: 2,5
+        let upsert_keys = UInt32Array::from(vec![2, 5]);
+        let upsert_values = UInt32Array::from(vec![200, 500]);
+        let upsert_vecs = FixedSizeListArray::try_new_from_values(
+            Float32Array::from(vec![21.0, 22.0, 23.0, 24.0, 25.0, 26.0, 27.0, 28.0]),
+            4,
+        )
+        .unwrap();
+
+        let upsert_batch = RecordBatch::try_new(
+            schema.clone(),
+            vec![
+                Arc::new(upsert_keys),
+                Arc::new(upsert_values),
+                Arc::new(upsert_vecs),
+            ],
+        )
+        .unwrap();
+
+        let upsert_stream = RecordBatchStreamAdapter::new(
+            schema.clone(),
+            futures::stream::once(async { Ok(upsert_batch) }).boxed(),
+        );
+
+        let (updated_dataset, _stats) =
+            MergeInsertBuilder::try_new(Arc::new(dataset), vec!["key".to_string()])
+                .unwrap()
+                .when_matched(WhenMatched::UpdateAll)
+                .when_not_matched(WhenNotMatched::DoNothing)
+                .when_not_matched_by_source(WhenNotMatchedBySource::Keep)
+                .try_build()
+                .unwrap()
+                .execute(Box::pin(upsert_stream))
+                .await
+                .unwrap();
+
+        let fragments = updated_dataset.get_fragments();
+        assert_eq!(fragments.len(), 3);
+
+        let updated_indices = updated_dataset.load_indices().await.unwrap();
+        let updated_value_index = updated_indices
+            .iter()
+            .find(|idx| idx.name == "value_idx")
+            .unwrap();
+        let updated_vec_index = updated_indices
+            .iter()
+            .find(|idx| idx.name == "vec_idx")
+            .unwrap();
+
+        // the frag bitmap of the index for 'value' field should not been updated
+        let value_bitmap = updated_value_index.fragment_bitmap.as_ref().unwrap();
+        assert_eq!(value_bitmap.len(), 2);
+        assert!(value_bitmap.contains(0));
+        assert!(value_bitmap.contains(1));
+
+        let vec_bitmap = updated_vec_index.fragment_bitmap.as_ref().unwrap();
+        assert_eq!(vec_bitmap.len(), 2);
+        assert!(vec_bitmap.contains(0));
+        assert!(vec_bitmap.contains(1));
+    }
+
+    #[tokio::test]
+    async fn test_sub_schema_upsert_fragment_bitmap() {
```

Comment:
Eventually, I'd like see a lot more tests cases for this. Unfortunately, right now the ratio of lines of code to useful tests cases is pretty high. I'd suggest thinking about how you can parameterize the tests to cover more useful situations. For example, you might use the same original test data, and very the merge insert parameters, input data, and index states.